### DISCUSSION TITLE
How to use vkCmdSetVertexInputEXT

### DISCUSSION THREAD

**Discussion by TareHimself** [terrafx/terrafx.interop.vulkan#64] [Q&A] [answered]
How do I call vkCmdSetVertexInputEXT, 
in c++ I can just do
`vkCmdSetVertexInputEXT(cmd,...params);`
 But in c# I cannot find a function to call

**Comment by tannergooding** [comment]
Extensions and many other APIs are not guaranteed exports and may not even be optimal to call via the export if they exist as one.

Rather, such APIs should be queried via `Command Function Pointers` instead, as per `46.4` in https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap46.html#extendingvulkan-extensions

TerraFX.Interop.Vulkan provides several `*ManualImports` types that contain function pointers to hold the resolved extensions for a given type, such as this one for `vkCmdSetVertexInputEXT`: https://source.terrafx.dev/#TerraFX.Interop.Vulkan/Vulkan/vulkan/vulkan_core/VkCommandBufferManualImports.cs,2aad28c9e5ec01cf

This helps ensure correctness by only exposing guaranteed exports and pushing developers to correctly enable features and resolve the non-static exports as part of initialization. This can also lead to better overall performance via bypassing the additional checks/queries that might be done for a general entry point that cannot be aware of whether the appropriate setup has taken place and thus must do additional validation or dispatch.

**Comment by TareHimself** [comment]
Thank you, but could I have an example of how to load and call a function using the c# bindings ?

**Reply by tannergooding** [reply, marked as answer]
```csharp
_vkCommandBufferManualImports.vkCmdSetVertexInputEXT = (delegate* unmanaged<VkCommandBuffer, uint, VkVertexInputBindingDescription2EXT*, uint, VkVertexInputAttributeDescription2EXT*, void>)vkGetDeviceProcAddr(vkDevice, (sbyte*)"vkCmdSetVertexInputEXT"u8.GetPointerUnsafe());
```

That is, you just use `vkGetDeviceProcAddr` to resolve the method for the specific device as detailed by https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetDeviceProcAddr.html

This can avoid dispatch overhead and generally be faster than had you resolved it otherwise; noting that `vkGetDeviceProcAddr` is valid for `device` or sub-objects created from `device` such as a `VkQueue` or `VkCommandBuffer`. 

For other imports, such as those taking a `VkInstance` you instead want `vkGetInstanceProcAddr`. This function pointer is also only usable with `instance` of sub-objects created from `instance` such as `VkPhysicalDevice`, `VkDevice, `VkQueue`, or `VkCommandBuffer`. This means it would also resolve APIs like `vkCmdSetVertexInputEXT` but may have overhead as it would need to dispatch to the specific instance for the `VkDevice`.
